src/main/java/com/puppycrawl/tools/checkstyle/LocalizedMessage.java

```diff
@@ -0,0 +1,158 @@
+///////////////////////////////////////////////////////////////////////////////////////////////
+// checkstyle: Checks Java source code and other text files for adherence to a set of rules.
+// Copyright (C) 2001-2022 the original author or authors.
+//
+// This library is free software; you can redistribute it and/or
+// modify it under the terms of the GNU Lesser General Public
+// License as published by the Free Software Foundation; either
+// version 2.1 of the License, or (at your option) any later version.
+//
+// This library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+// Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public
+// License along with this library; if not, write to the Free Software
+// Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
+///////////////////////////////////////////////////////////////////////////////////////////////
+
+package com.puppycrawl.tools.checkstyle;
+
+import java.io.IOException;
+import java.io.InputStreamReader;
+import java.io.Reader;
+import java.net.URL;
+import java.net.URLConnection;
+import java.nio.charset.StandardCharsets;
+import java.text.MessageFormat;
+import java.util.Arrays;
+import java.util.Locale;
+import java.util.MissingResourceException;
+import java.util.PropertyResourceBundle;
+import java.util.ResourceBundle;
+import java.util.ResourceBundle.Control;
+
+/**
+ * Represents a message that can be localised. The translations come from
+ * message.properties files. The underlying implementation uses
+ * java.text.MessageFormat.
+ */
+public class LocalizedMessage {
+
+    /** Name of the resource bundle to get messages from. **/
+    private final String bundle;
+
+    /**
+     * The locale to localise messages to.
+     **/
+    private final Locale locale;
+
+    /** Class of the source for this message. */
+    private final Class<?> sourceClass;
+
+    /**
+     * Key for the message format.
+     **/
+    private final String key;
+
+    /**
+     * Arguments for MessageFormat.
+     */
+    private final Object[] args;
```

Comment:
Why did we use `Object[]` instead of `String[]`?

Reply:
The code in this class is centrialized from DefaultLogger and Violation.
While DefaultLogger used `String[]`, Violation has its input in its constructor to be `Object[]`. I kept this class `Object[]` since I didn't think I could easily have Violation convert `Object[]` to `String[]`.
The method that receives this `args` just takes `Object obj`.